cmd/deployment-check/deployment.go

```diff
@@ -0,0 +1,703 @@
+		Template:        podTemplateSpec,
+	}
+
+	// Define the k8s deployment.
+	deployment.ObjectMeta.Name = checkDeploymentName
+	deployment.ObjectMeta.Namespace = checkNamespace
+
+	// Add the deployment spec to the deployment.
+	deployment.Spec = deploySpec
+
+	return deployment
+}
+
+// DeploymentResult represents the results from a createDeployment and updateDeployment calls.
+type DeploymentResult struct {
+	Deployment *v1.Deployment
+	Err        error
+}
+
+// createDeployment creates a deployment in the cluster with a given deployment specification.
+func createDeployment(deploymentConfig *v1.Deployment) chan DeploymentResult {
+
+	createChan := make(chan DeploymentResult)
+
+	go func() {
+		log.Infoln("Creating deployment in cluster with name:", deploymentConfig.Name)
+
+		defer close(createChan)
+
+		result := DeploymentResult{}
+
+		deployment, err := client.AppsV1().Deployments(checkNamespace).Create(deploymentConfig)
+		if err != nil {
+			log.Infoln("Failed to create a deployment in the cluster:", err)
+			result.Err = err
+			createChan <- result
+			return
+		}
+
+		for {
+			log.Infoln("Watching for deployment to exist.")
+
+			// Watch that it is up.
+			watch, err := client.AppsV1().Deployments(checkNamespace).Watch(metav1.ListOptions{
+				Watch:         true,
+				FieldSelector: "metadata.name=" + deployment.Name,
+				// LabelSelector: defaultLabelKey + "=" + defaultLabelValueBase + strconv.Itoa(int(now.Unix())),
+			})
+			if err != nil {
+				result.Err = err
+				createChan <- result
+				return
+			}
+
+			// There can be 2 events here: Available = True status update from deployment or Context timeout.
+			for event := range watch.ResultChan() { // Watch for deployment events.
+
+				d, ok := event.Object.(*v1.Deployment)
+				if !ok { // Skip the event if it cannot be casted as a v1.Deployment.
+					log.Infoln("Got a watch event for a non-deployment object -- ignoring.")
+					continue
+				}
+
+				log.Debugln("Received an event watching for deployment changes:", d.Name, "got event", event.Type)
+
+				// Look at the status conditions for the deployment object;
+				// we want it to be reporting Available = True.
+				if deploymentAvailable(d) {
+					result.Deployment = d
+					createChan <- result
+					return
+				}
+
+				// If the context has expired, exit.
+				select {
+				case <-ctx.Done(): // Watch for a context cancellation.
+					log.Errorln("Context expired while waiting for deployment to create.")
+					err = cleanUp() // Clean up the deployment.
+					if err != nil {
+						result.Err = errors.New("failed to clean up properly: " + err.Error())
+					}
+					createChan <- result
+					return
+				default:
+				}
+			}
+
+			// Stop the watch on each loop because we will create a new one.
+			watch.Stop()
+		}
+	}()
+
+	return createChan
+}
+
+// createContainerConfig creates a container resource spec and returns it.
+func createContainerConfig(imageURL string) corev1.Container {
```

Comment:
this is a good candidate for a unit test

Reply:
basic tests for config creation have been added